### PR TITLE
Added /build/three.es2015.js target for "jsnext:main"

### PR DESCRIPTION
**Pull request by cecilemuller**
Rollup builds both the `umd` and a new `es` bundle now

### CODE REVIEW

**Comment by mrdoob** [comment]
Is `foo.es2015.js` a standard or something? Otherwise, how about `three.jsnext` instead?

**Comment by cecilemuller** [comment]
As long as it's linked from package.json, it can be anything you want, so sure let's call it `three.jsnext` instead :smiley:

**Comment by mrdoob** [comment]
Hmm... Is there any good reading about what jsnext is? 😶

**Comment by cecilemuller** [comment]
Sure, here's the [documentation from Rollup](https://github.com/rollup/rollup/wiki/jsnext:main) about it.

**Comment by GGAlanSmithee** [comment]
[Here](https://github.com/jsforum/jsforum/issues/5) is a lengthy discussion on `jsnext:main`. Not everyone is for it, but it is gaining some traction in the echosystem.

Also, I agree with @mrdoob that `.es2015` might not be a good name, since `jsnext` points to the latest iteration of ecmascript.

**Comment by mrdoob** [comment]
My question is... Would it be weird if it was just `three.jsnext` instead of `three.jsnext.js`?

**Comment by GGAlanSmithee** [comment]
Ah sorry, missed that.. IMO that doesn't really make sense since the syntax is stull 100% pure js. In react for example, some people use the extension jsx, but that is different IMO since react uses jsxml, which is not pure js. 

It is a matter of opinion though, I guess.

**Comment by cecilemuller** [comment]
I'd personally prefer using `.js` (because it is real Javascript), although I've seen `.mjs` used informally for "exports an ES module" (but it doesn't seem to have much traction) whereas I never encountered a `.jsnext` file.

To compare, here's searching files on Github that use `import`, by file extension:
 - [.js](https://github.com/search?utf8=%E2%9C%93&q=import+extension%3Ajs&type=Code&ref=searchresults): 3,984,568 results
 - [.mjs](https://github.com/search?utf8=%E2%9C%93&q=import+extension%3Amjs&type=Code&ref=searchresults): 19 results
 - [.jsnext](https://github.com/search?utf8=%E2%9C%93&q=import+extension%3Ajsnext&type=Code&ref=searchresults): 0 results

**Comment by vanruesc** [comment]
How about calling the jsnext bundle `three.js` and the backwards-compatibility bundle `three.es5.js`? The latter will remain ES5 code until it gets removed some day, so that's a constant.

**Comment by cecilemuller** [comment]
Having the new bundle not named `three.js` lets users opt-in to an ES Modules codebase when they're ready, without breaking existing applications that don't use npm to get the built files.

**Comment by mrdoob** [comment]
`three.next.js`? 😅

**Comment by cecilemuller** [comment]
It's renamed :smiley:

File: package.json [hunk 1]
```diff
@@ -3,11 +3,13 @@
   "version": "0.80.1",
   "description": "JavaScript 3D library",
   "main": "build/three.js",
+  "jsnext:main": "build/three.jsnext.js",
```
Comment (mrdoob) [inline package.json]:
```
"jsnext:main": "build/three.next.js",
```

**Comment by cecilemuller** [comment]
Good catch :)

**Comment by mrdoob** [comment]
Sorry to keep going back and forth on this one.

Thinking about the future... Is the expectation that `three.next.js` would always be using the latest es*?

**Comment by cecilemuller** [comment]
I think the only expectation is just that it can be imported as an ES module (to be able to use tree shaking), nothing binding about using other es features.

**Comment by mrdoob** [comment]
Ah... Should it be named `three.module.js` then?

**Comment by mrdoob** [comment]
I mean, we're adding a new file to `/builds`. That hasn't happened in 6 years. Just want to make sure users aren't confused.

**Comment by mrdoob** [comment]
Also, I thought the idea was that people were going to include `src/Three.js`? Isn't that the same?

**Comment by cecilemuller** [comment]
Making `jsnext:main` point to `src/Three.js` used to be the plan, but there is a required [glsl() transform](https://github.com/mrdoob/three.js/blob/r80/rollup.config.js#L10) that prevents using it directly.

Also, the [`src` folder isn't included](https://github.com/mrdoob/three.js/blob/r80/.npmignore) in the NPM package.

But if you prefer changing the codebase to not need that glsl transform, and adding the `src` folder on NPM, sure let's do that instead.

**Comment by mrdoob** [comment]
> Also, the [`src` folder isn't included](https://github.com/mrdoob/three.js/blob/r80/.npmignore) in the NPM package.

Oh...? https://github.com/mrdoob/three.js/blob/dev/package.json#L12

**Comment by cecilemuller** [comment]
Just tried a fresh install and I was wrong, .npmignore isn't preventing src from being included, great :)

Still leaves the glsl issue though.

**Comment by magnonellie** [comment]
Is this dead? It would be really nice to have it... :disappointed:

**Comment by mrdoob** [comment]
@cecilemuller What do you think about naming the file `three.modules.js`?

**Comment by cecilemuller** [comment]
@mrdoob That name sounds fine 😃

Ok so I renamed the target in the fork, updated Rollup to the latest version, and re-opened the PR.

For reference, here's also an example app that imports the new module bundle (to make sure it actually works for end users): https://github.com/wildpeaks/example-threejs-esmodule

**Comment by mrdoob** [comment]
Sweet!

**Comment by mrdoob** [comment]
Thanks!

**Comment by cecilemuller** [comment]
One caveat: I just tried with [a less trivial test application](https://wildpeaks.github.io/example-threejs-esmodule/) and one small issue remains.

In the original Rollup config, there was this block added as `outro` (which I initially left untouched because I only wanted to add the second build target):

	Object.defineProperty( exports, 'AudioContext', {
		get: function () {
			return exports.getAudioContext();
		}
	});

However it's appended as-is like text to the bundle, so at runtime it doesn't know what "exports" is.
So I removed it (because it's just aliasing another function anyway ? apps using `THREE.AudioContext` could directly use `THREE.getAudioContext()` instead) and re-built the modules bundle, and now it renders fine :)

However I'm still trying to publish that change to the fork because right now I'm fighting the UI to get the fork resynced first (because usually I never need to keep a fork after it's merged and just delete it).

**Comment by cecilemuller** [comment]
I opened a separated pull request for continuing that discussion:
https://github.com/mrdoob/three.js/pull/9901

**Comment by cfebs** [comment]
@cecilemuller - I am trying the builds in the test application and noticed that bundle size does not change as drastically based on which modules you import.

```
# in application.js
import {Scene, PerspectiveCamera, WebGLRenderer, BoxGeometry, MeshBasicMaterial, Mesh} from 'three';

# after build
$ wc -c docs/bundle.iife.js 
759212 docs/bundle.iife.js
```

vs. something that should produce a vastly smaller bundle (I was assuming?)

```
# in applicationl.js
import {Bone} from 'three';

# nothing else in the file
$ wc -c ./docs/bundle.iife.js 
758518 ./docs/bundle.iife.js
```

The production minified lib gziped is about 124k, these sizes are 112k just importing a small amt of modules.

I was able to produce much smaller bundle sizes by working with imports that look like this: https://gist.github.com/cfebs/f2f3ec134730b1cd44269cc04db8d552

After uglify + gzip, it's down to 12k . This is more of what i would expect.

I don't really understand why this is. Maybe the my method is too naive?

**Comment by looeee** [comment]
@cfebs shouldn't you be comparing those files to the unminified build? Which is currently 979k.

**Comment by mrdoob** [comment]
@cfebs I did some research on this some weeks ago. Seems like Rollup can't tree-shake `Object.assign( class.prototype, { ... } )`. I wonder if it's able to tree-shake other approaches though.

https://github.com/mrdoob/model-tag/issues/1

**Comment by cfebs** [comment]
Sure here is the comparison in a nicer layout

| method  | modules  | unminified  | minified  | minified+gzip |
|---|---|---|---|---|
| production dist  | all  | 979k  | 491k | 124k |
| test app rollup  | Math, Bone | 758k  | 438k  | 112k  |
| using src refs  | Math, Bone  | 75k  | 46k  | 12k  |

If more extensive code samples are needed i can organize further, i'm a little scatterbrained w/ these builds.

Also if moving to new ticket is better instead of dead PR i can do that as well.

**Comment by Rich-Harris** [comment]
> Seems like Rollup can't tree-shake `Object.assign( Mesh.prototype, { ... } )`. I wonder if it's able to tree-shake other approaches though.

Yep, it's not (yet) that sophisticated. It has a concept of pure functions – i.e. it knows that if you call, say, `var keys = Object.keys(Mesh.prototype)`, you're not creating any side-effects that need inclusion independently of how `keys` is then used. The next level would be for it to understand that `Object.assign(...)` only has side-effects that pertain to its first argument, and that if `Mesh` isn't used elsewhere then the statement can be excluded. It's on our wishlist!

**Comment by looeee** [comment]
@Rich-Harris Would it work if

```js
Object.assign( Mesh.prototype, {

    constructor: Mesh,
    isMesh: true,
    setDrawMode: function ( ) { .. .},
    ....
 } )
```
  was changed to 

```js
Mesh.prototype.constructor = Mesh;

Mesh.prototype.isMesh = true;

Mesh.prototype.setDrawMode = function () { ... }

...
```

**Comment by Rich-Harris** [comment]
@looeee yes, in theory that should work – see [this](http://rollupjs.org/?version=0.36.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyY2xhc3MlMjBVbnVzZWQlMjAlN0IlN0QlM0IlNUNuJTVDbk9iamVjdC5hc3NpZ24oJTIwVW51c2VkLnByb3RvdHlwZSUyQyUyMCU3QiU1Q24lMjAlMjBmb28lMjAoKSUyMCU3QiU3RCU1Q24lN0QpJTNCJTIyJTdEJTVEJTdE) versus [this](http://rollupjs.org/?version=0.36.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyY2xhc3MlMjBVbnVzZWQlMjAlN0IlN0QlM0IlNUNuJTVDblVudXNlZC5wcm90b3R5cGUuZm9vJTIwJTNEJTIwZnVuY3Rpb24lMjBmb28lMjAoKSUyMCU3QiU3RCUzQiUyMiU3RCU1RCU3RA==), in the second example `Unused` is excluded.

**Comment by looeee** [comment]
@mrdoob, @tschw  as I've been going over the code over the last while, it seems use of Object.assign was added fairly recently (mainly in R77 I think). Was there a reason for this?

**Comment by mrdoob** [comment]
> @mrdoob, @tschw  as I've been going over the code over the last while, it seems use of Object.assign was added fairly recently (mainly in R77 I think). Was there a reason for this? 

Interesting... I think we changed it because we *felt* `Object.assign()` was more performant. It also *felt* more compact. The resulting minified file was smaller, but after gzip there wasn't much difference.

I guess I'll try converting some classes back and see how rollup deals with it 😊

**Comment by mrdoob** [comment]
We also use `Object.create()` for inheritance though...

**Comment by Rich-Harris** [comment]
That's okay – Rollup knows that `Object.create` doesn't have side-effects. [Demo here](http://rollupjs.org/?version=0.37.0&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyY2xhc3MlMjBVbnVzZWQlMjAlN0IlN0QlNUNuJTVDbmNvbnN0JTIwdW51c2VkJTIwJTNEJTIwT2JqZWN0LmNyZWF0ZSglMjBVbnVzZWQucHJvdG90eXBlJTIwKSUzQiU1Q24lNUNuZXhwb3J0JTIwZGVmYXVsdCUyMDQyJTNCJTIyJTdEJTVEJTdE)

**Comment by looeee** [comment]
@Rich-Harris you beat me to it :) 

@mrdoob here's an extended rollup example

Inheritance using [Object.assign](http://rollupjs.org/?version=0.36.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJtb2R1bGVJZCUyMiUzQSUyMiUyMiU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlMkYlMkZCJTIwd2lsbCUyMGluaGVyaXQlMjBmcm9tJTIwQSU1Q24lNUNuZnVuY3Rpb24lMjBBKCklN0IlNUNuJTIwJTIwdGhpcy52YXJBJTIwJTNEJTIwJ2EnJTNCJTVDbiU3RCU1Q24lNUNuQS5wcm90b3R5cGUlMjAlM0QlMjAlN0IlNUNuJTIwJTIwZG9Tb21ldGhpbmdGcm9tQSUyMCUzQSUyMGZ1bmN0aW9uKCklN0IlNUNuJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ1NvbWV0aGluZyUyMGZyb20lMjBBJTNBJTIwJyUyMCUyQiUyMHRoaXMudmFyQSklM0IlNUNuJTIwJTIwJTdEJTVDbiU3RCUzQiU1Q24lNUNuZnVuY3Rpb24lMjBCKCklN0IlNUNuJTIwJTIwQS5jYWxsKHRoaXMpJTNCJTVDbiUyMCUyMHRoaXMudmFyQiUyMCUzRCUyMGIlM0IlNUNuJTdEJTVDbkIucHJvdG90eXBlJTIwJTNEJTIwT2JqZWN0LmFzc2lnbihBLnByb3RvdHlwZSUyQyUyMCU3QiU1Q24lMjAlMjBkb1NvbWV0aGluZ0Zyb21CJTIwJTNBJTIwZnVuY3Rpb24oKSU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygnU29tZXRoaW5nJTIwZnJvbSUyMEIlM0ElMjAnJTIwJTJCJTIwdGhpcy52YXJCKSUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEKSUzQiU1Q25CLnByb3RvdHlwZS5jb25zdHJ1Y3RvciUyMCUzRCUyMEIlM0IlNUNuJTVDbnZhciUyMGElMjAlM0QlMjBuZXclMjBBKCklM0IlNUNuJTIyJTdEJTVEJTdE), no tree shaking occurs.

Inheritance using [Object.create](http://rollupjs.org/?version=0.36.4&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJtb2R1bGVJZCUyMiUzQSUyMiUyMiU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlMkYlMkZCJTIwd2lsbCUyMGluaGVyaXQlMjBmcm9tJTIwQSU1Q24lNUNuZnVuY3Rpb24lMjBBKCklN0IlNUNuJTIwJTIwdGhpcy52YXJBJTIwJTNEJTIwJ2EnJTNCJTVDbiU3RCU1Q24lNUNuQS5wcm90b3R5cGUlMjAlM0QlMjAlN0IlNUNuJTIwJTIwZG9Tb21ldGhpbmdGcm9tQSUyMCUzQSUyMGZ1bmN0aW9uKCklN0IlNUNuJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ1NvbWV0aGluZyUyMGZyb20lMjBBJTNBJTIwJyUyMCUyQiUyMHRoaXMudmFyQSklM0IlNUNuJTIwJTIwJTdEJTVDbiU3RCUzQiU1Q24lNUNuZnVuY3Rpb24lMjBCKCklN0IlNUNuJTIwJTIwQS5jYWxsKHRoaXMpJTNCJTVDbiUyMCUyMHRoaXMudmFyQiUyMCUzRCUyMGIlM0IlNUNuJTdEJTVDbkIucHJvdG90eXBlJTIwJTNEJTIwT2JqZWN0LmNyZWF0ZShBLnByb3RvdHlwZSUyQyUyMCU3QiU1Q24lMjAlMjBkb1NvbWV0aGluZ0Zyb21CJTIwJTNBJTIwZnVuY3Rpb24oKSU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygnU29tZXRoaW5nJTIwZnJvbSUyMEIlM0ElMjAnJTIwJTJCJTIwdGhpcy52YXJCKSUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEKSUzQiU1Q25CLnByb3RvdHlwZS5jb25zdHJ1Y3RvciUyMCUzRCUyMEIlM0IlNUNuJTVDbnZhciUyMGElMjAlM0QlMjBuZXclMjBBKCklM0IlNUNuJTIyJTdEJTVEJTdE), tree shaking does occur.

As I understand it, the advantage of Object.assign over Object.create is that you can do 'multiple inheritance:

```js
var C = Object.assign( X.prototype, Y.prototype, {
    ...
});

```
Also it looks neater :) 

But this might be a good enough reason not to use it for now.

**Comment by mrdoob** [comment]
> That's okay – Rollup knows that `Object.create` doesn't have side-effects. [Demo here](http://rollupjs.org/?version=0.37.0&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyY2xhc3MlMjBVbnVzZWQlMjAlN0IlN0QlNUNuJTVDbmNvbnN0JTIwdW51c2VkJTIwJTNEJTIwT2JqZWN0LmNyZWF0ZSglMjBVbnVzZWQucHJvdG90eXBlJTIwKSUzQiU1Q24lNUNuZXhwb3J0JTIwZGVmYXVsdCUyMDQyJTNCJTIyJTdEJTVEJTdE)


Hmm, [demo updated with three.js style](http://rollupjs.org/?version=0.37.0&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZnVuY3Rpb24lMjBVbnVzZWQoKSUyMCU3QiU3RCU1Q25mdW5jdGlvbiUyMFVudXNlZDIoKSUyMCU3QiU3RCU1Q25VbnVzZWQyLnByb3RvdHlwZSUyMCUzRCUyME9iamVjdC5jcmVhdGUoJTIwVW51c2VkLnByb3RvdHlwZSUyMCklM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjA0MiUzQiUyMiU3RCU1RCU3RA==).

I tried yesterday porting all `THREE.Audio` classes to non-`Object.assign()` and I couldn't get it to get tree-shaked on `model-tag`.

I guess I'll give it another go on the r84 cycle ✌️

**Comment by mrdoob** [comment]
@Rich-Harris Let us know when Rollup can handle `Object.assign()`. The tree-shaking should then be massive! 😊 [demo with Object.assign()](http://rollupjs.org/?version=0.37.0&shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZnVuY3Rpb24lMjBVbnVzZWQoKSUyMCU3QiU3RCU1Q25VbnVzZWQucHJvdG90eXBlJTIwJTNEJTIwJTdCJTVDbiUyMCUyMHNldCUzQSUyMGZ1bmN0aW9uJTIwKCklMjAlN0IlMjBjb25zb2xlLmxvZyglMjAncG90b3RlJyUyMCklMjAlN0QlMkMlNUNuJTIwJTIwc2V0MiUzQSUyMGZ1bmN0aW9uJTIwKCklMjAlN0IlMjBjb25zb2xlLmxvZyglMjAncGFwYSclMjApJTIwJTdEJTVDbiU3RCUzQiU1Q25mdW5jdGlvbiUyMFVudXNlZDIoKSUyMCU3QiU1Q24lMjAlMjBVbnVzZWQuY2FsbCglMjB0aGlzJTIwKSUzQiU1Q24lN0QlNUNuT2JqZWN0LmFzc2lnbiglMjBVbnVzZWQyLnByb3RvdHlwZSUyQyUyMFVudXNlZC5wcm90b3R5cGUlMjApJTNCJTVDbiU1Q252YXIlMjB0ZXN0JTIwJTNEJTIwbmV3JTIwVW51c2VkKCklM0IlNUNudGVzdC5zZXQoKSUzQiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwNDIlM0IlMjIlN0QlNUQlN0Q=)

**Pull request merged**